### PR TITLE
Disambiguate brace spacing

### PR DESCRIPTION
**Pull request by schwagogo**
Pick one rule as to whether or not to use spaces inside {} for each way that Ruby uses braces.

At present some people do and some do not put spaces inside braces used for blocks or hash literals.
We should have a single rule that can be followed by automated formatters (which means not treating block and hash literals differently.)

I don't think anyone uses or wants spaces inside #{}, so let's just remove that ambiguity.
### tl;dr of proposed changes
- `ary.each { |x| x.do_thingy }` good; `ary.each {|x| x.do_thingy}` bad
- `"this is a #{adjective} string"` good; `"this is a #{ adjective } string"` bad

Spaces around brackets when used as a hash or block, no spaces when used for interpolation.

To see the changes in context, search for "deserve a bit of clarification" in https://github.com/IndieGoGo/ruby-style-guide/tree/schwa/disambiguate-brace-spacing.

### CODE REVIEW

**Comment by jacobaweiss** [comment]
omg yis yis yis :+1: :100: :icecream: :muscle: 

not using spaces in block literals & hashes mentally clutters the area for me, i'm into this.

**Comment by edgeworthigg** [comment]
What about blocks inside embedded expressions, e.g. `"#{this or that}"`?  I find the spaces help there.

**Comment by gogojared** [comment]
I don't think this helps readability, and tend to judge the spaces inside expressions based on visual density of surrounding code. If you're trying to lint though, alright, +0. Feel like I'm going to violate this one accidentally a lot.

**Comment by godylango** [comment]
:+1: I am on board with this change.

**Comment by jenmei** [comment]
If we are automating it, then I'd prefer multiline. I'd be okay for exceptions for short one, like the example, but if we choose only one format, I'd say multiline:

some_var = {
  one: 1,
  two: 2
}

I'd also be fine w/ a trailing comma:

some_var = {
  one: 1,
  two: 2,
}

Also, I'm for letting people use their judgement w/ string interpolation.

**Comment by schwagogo** [comment]
@jenmei This PR isn't about single-line or multiline. It's about `{foo: bar}` vs. `{ foo: bar }` and `"foo#{bar}baz"` vs. `"foo#{ bar }baz"`.

**Comment by mLuby** [comment]
I don't have an opinion on the spacing, but I'm strongly in favor of style conformity via auto-linting. So whichever makes doing that easier.

**Comment by jamesgogo** [comment]
{ james: &#128077; }

**Comment by jacobaweiss** [comment]
@schwagogo in regards to auto-linting, I'd happy take the lead / work with you on getting [hound](https://houndci.com/) set up. It runs rubocop on your diff, and leaves inline comments. It'd probably be noisy at first, but as we get better at it, style would probably become much more consistent.

**Comment by jacobaweiss** [comment]
also 

### tl;dr of proposed changes
- `ary.each { |x| x.do_thingy }` good; `ary.each {|x| x.do_thingy}` bad
- `"this is a #{adjective} string"` good; `"this is a #{ adjective } string"` bad
Spaces around brackets when used as a hash or block, no spaces when used for interpolation.

**Comment by schwagogo** [comment]
@jacobaweiss good call on illustrating the changes here. Probably a good idea for future style guide PRs.
Re actually running rubocop, yes. I'm thinking we should do some rounds of running it manually and letting it autofix before we try the PR comments.

**Comment by schwagogo** [comment]
Oh, FWIW, when I said "automated formatters" I was thinking of RubyMine. My real agenda here is to get RubyMine users to set their Ruby formatter the same way regarding spaces inside braces. But rubocop is still a good idea.

**Comment by schwagogo** [comment]
I count 4 :+1: , 1 :-1: 2 +0 so far.

**Comment by gogowill** [comment]
+1.

**Comment by npj** [comment]
+1 for spaces

**Comment by gogonathan** [comment]
My current natural style is this:

I'm for spaces in blocks:
{ |var| do_this(var) }

but no spaces for one-line hashes:
{foo: 'bar', baz: 'bah'}

I find the existence or lack of a space to be a useful subliminal hint
about whether I'm looking at a hash or a block.

On Tue, Jan 27, 2015 at 5:54 PM, Dave Schweisguth <notifications@github.com>
wrote:

> I count 4 [image: :+1:] , 1 [image: :-1:] 2 +0 so far.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/IndieGoGo/ruby-style-guide/pull/4#issuecomment-71767498>
> .
>



-- 
Nathan Wilmes
Senior Developer
Indiegogo
*866.641.4646*
*@gogo_nathan*

**Comment by GoGoAmanda** [comment]
+1 LGTM :shipit:

**Comment by IndieKelly** [comment]
+0 to actual standard.
+1 to being able to use rubymine or an formatter to fix it.

**Comment by paulz** [comment]
+1 what Nathan said about hashes. I would even drop spaces in blocks, but ok with them

**Comment by judygogo** [comment]
{ judy: :+1: }

**Comment by gogodan** [comment]
+1

**Comment by jenmei** [comment]
@schwagogo, just to clarify, I get this is about spaces and not newlines. My comment above was specifically addressing auto-formatting. I would prefer an auto-formatter to format literal hashes in multi-line format, unless we could use complex rules (e.g. multi-line hash literals unless they are short and appear inside array literals). E.g.

```ruby
# multi-line if it's just one hash
professor = {
  first: 'Albus',
  last: 'Dumbledore'
}

# but if hashes are short and they appear in an array, one line is great
professors = [
  { first: 'Albus', last: 'Dumbledore'},
  { first: 'Minerva', last: 'McGonagall'}
]

# but if they are complex, then multiline for arrays of hashes

professors = [
  { 
    first: 'Albus', 
    last: 'Dumbledore',
    school: 'Hogwarts', 
    titles: [
      'Headmaster',
      'Professor'
    ]
  },
  { 
    first: 'Minerva', 
    last: 'McGonagall',
    school: 'Hogwarts', 
    titles: [
      'Professor',
      'Head of Gryffindor'
    ]
  }
]
```

I realize it's common to configure auto-formatting to not insert line breaks, but I frequently find myself wanting auto line-break insertion.
  
That said, when hash literals appear on one line, spaces are better than no spaces.
  
More commonly, this would come up for me w/ one-line blocks, and I definitely prefer spaces there:
  
  ```ruby
  collection.sort { |a, b| a <=> b }
  ```
  
So I'm okay w/ this pull request with some caveats which may appear in a future pull request.

**Comment by valgogo** [comment]
How shall empty hash literals be handled? For example, I've seen empty hashes formatted in two different ways in our code base:
* `{ }`
* `{}`

My preference is for the first format as it matches the desired spacing for non-empty hash literals. The same could be applied to empty arrays: `[ ]`

I think we should formalize the convention, as I've been confused which one is preferred.

**Comment by schwagogo** [comment]
@valgogo the current code standard doesn't specifically address that, but uses `{}` several times and does not use `{ }`. From code I've seen (both in and out of the company) everyone uses `{}`. If anyone wants to change that in the code standard it would be a new PR -- don't want to move the goalposts on this one.

**Comment by schwagogo** [comment]
OK, voting closed.
+1: @jacobaweiss @jenmei @godylango @mLuby @jamesgogo @gogowill @npj @GoGoAmanda @judygogo @gogodan 
0: @gogojared @gogonathan @IndieKelly @paulz
-1: @gogonathan, @paulz

The ayes have it. I'll merge, look in to what RubyMine does a bit and mail the team.
